File: scripts/appsmith-import.py
#!/usr/bin/env python3
"""
Appsmith Import Automation via Playwright (Python)
Imports appsmith-import.json into Appsmith workspace
"""

import time
from pathlib import Path
from playwright.sync_api import sync_playwright, TimeoutError as PlaywrightTimeout

APPSMITH_URL = 'https://appsmith-production-f331.up.railway.app'
EMAIL = 'ewertonfm00@gmail.com'
PASSWORD = 'Mll031119'
JSON_FILE = r'Z:\My Folders\Projetos - Claude Code\AIOX - Machine Learning\portal\appsmith-import.json'
SCREENSHOTS_DIR = r'Z:\My Folders\Projetos - Claude Code\AIOX - Machine Learning\scripts\screenshots'


def ss(page, name):
    Path(SCREENSHOTS_DIR).mkdir(parents=True, exist_ok=True)
    path = str(Path(SCREENSHOTS_DIR) / f"{name}.png")
    page.screenshot(path=path, full_page=True)
    print(f"  [screenshot] {name}.png")


def dump_page(page, label=""):
    """Print current URL and visible button texts"""
    print(f"  URL ({label}): {page.url}")
    try:
        texts = page.evaluate("""() => {
            const els = document.querySelectorAll('button, a, [role="menuitem"], [role="button"], li');
            return Array.from(els)
                .map(el => el.textContent?.trim())
                .filter(t => t && t.length > 0 && t.length < 60)
                .slice(0, 40);
        }""")
        print(f"  Elements: {texts}")
    except Exception as e:
        print(f"  (could not dump elements: {e})")


def main():
    print("=" * 60)
    print("Appsmith Import Automation")
    print("=" * 60)
    print(f"URL: {APPSMITH_URL}")
    print(f"File: {JSON_FILE}")
    print(f"File exists: {Path(JSON_FILE).exists()}")

    with sync_playwright() as p:
        browser = p.chromium.launch(
            headless=False,
            slow_mo=200,
        )
        context = browser.new_context(viewport={'width': 1440, 'height': 900})
        page = context.new_page()

        try:
            # ── STEP 1: Go to login page ──
            print("\n[1] Navigating to login page...")
            page.goto(f"{APPSMITH_URL}/user/login", wait_until='domcontentloaded', timeout=30000)
            time.sleep(3)
            ss(page, '01-login')
            dump_page(page, 'login')

            # ── STEP 2: Fill login form ──
            print("\n[2] Filling login form...")

            # Wait for email input
            page.wait_for_selector('input[name="username"]', timeout=10000)
            page.fill('input[name="username"]', EMAIL)
            page.fill('input[name="password"]', PASSWORD)
            ss(page, '02-form-filled')

            # Click Sign In button
            page.click('button[type="submit"]')
            print("  Clicked Sign In")

            # Wait for redirect after login
            try:
                page.wait_for_url(lambda url: '/user/login' not in url, timeout=20000)
            except PlaywrightTimeout:
                pass
            time.sleep(3)
            ss(page, '03-after-login')
            dump_page(page, 'after-login')

            # ── STEP 3: Navigate to workspace home ──
            print("\n[3] At workspace...")
            # Check if we're on the applications page
            current_url = page.url
            print(f"  Current URL: {current_url}")

            # Wait for workspace to load
            time.sleep(2)
            ss(page, '04-workspace')
            dump_page(page, 'workspace')

            # ── STEP 4: Find Import option ──
            # In Appsmith, import is typically under a dropdown from the workspace
            # or from "Create New" -> "Import"
            print("\n[4] Looking for Import option...")

            # First try: look for a direct "Import" button
            import_visible = page.locator('button:has-text("Import"), span:has-text("Import")').first
            try:
                if import_visible.is_visible(timeout=3000):
                    print("  Found direct Import button!")
                    import_visible.click()
                    time.sleep(2)
                    ss(page, '05-import-clicked')
                else:
                    raise Exception("not visible")
            except:
                print("  Direct Import not visible, trying Create New dropdown...")
                # Try "Create New" button
                try:
                    page.click('button:has-text("Create New")', timeout=5000)
                    time.sleep(1)
                    ss(page, '05-create-new-menu')
                    dump_page(page, 'create-new-menu')
                except:
                    print("  Create New button not found either")
                    ss(page, '05-fallback')
                    dump_page(page, 'fallback')

            # ── STEP 5: Look for "Import from file" in dropdown/dialog ──
            print("\n[5] Looking for 'Import from file'...")
            time.sleep(1)

            # Try clicking "Import" in a dropdown menu
            try:
                page.click('text=Import', timeout=5000)
                time.sleep(1)
                ss(page, '06-import-option')
                dump_page(page, 'import-option')
            except:
                print("  Could not click 'Import' text")

            # ── STEP 6: Handle file upload ──
            print("\n[6] Looking for file input...")
            time.sleep(2)
            ss(page, '07-before-file')

            # Try "Import from file" link
            try:
                page.click('text=/import from file/i', timeout=4000)
                time.sleep(1)
                ss(page, '07b-after-import-from-file')
            except:
                pass

            # Look for file input (hidden or visible)
            file_input = page.locator('input[type="file"]').first
            try:
                file_input.wait_for(timeout=5000)
                print(f"  Found file input! Setting file: {JSON_FILE}")
                file_input.set_input_files(JSON_FILE)
                time.sleep(3)
                ss(page, '08-file-set')
                print("  File selected!")
            except Exception as e:
                print(f"  File input not found: {e}")
                ss(page, '08-no-file-input')
                dump_page(page, 'no-file-input')

            # ── STEP 7: Wait for import progress ──
            print("\n[7] Waiting for import to process...")
            time.sleep(5)
            ss(page, '09-processing')
            dump_page(page, 'processing')

            # ── STEP 8: Handle datasource reconnection ──
            print("\n[8] Checking for datasource dialog...")
            time.sleep(2)

            # Appsmith shows a "Reconnect Datasource" modal after import
            reconnect_visible = page.locator('text=/reconnect/i, text=/datasource/i').count()
            print(f"  Datasource/reconnect elements: {reconnect_visible}")
            ss(page, '10-datasource-check')

            if reconnect_visible > 0:
                print("  Found datasource reconnection dialog!")
                dump_page(page, 'datasource-dialog')

                # Try to find ML Portal DB
                try:
                    ml_db = page.locator('text="ML Portal DB"').first
                    if ml_db.is_visible(timeout=3000):
                        print("  ML Portal DB found!")
                        ml_db.click()
                        time.sleep(1)
                except:
                    pass

                # Look for action buttons
                for btn_text in ['Skip to App', 'Skip', 'Done', 'Finish', 'Continue']:
                    try:
                        btn = page.locator(f'button:has-text("{btn_text}")').first
                        if btn.is_visible(timeout=2000):
                            print(f"  Clicking '{btn_text}' button")
                            btn.click()
                            time.sleep(2)
                            break
                    except:
                        pass

                ss(page, '11-after-datasource')

            # ── STEP 9: Check final state ──
            print("\n[9] Final state...")
            time.sleep(4)
            ss(page, '12-final')
            dump_page(page, 'final')

            final_url = page.url
            print(f"\n  Final URL: {final_url}")

            # Check for expected pages
            for page_name in ['Projetos', 'Skills', 'Perfil', 'Fila', 'Numbers']:
                count = page.locator(f'text="{page_name}"').count()
                if count > 0:
                    print(f"  FOUND: '{page_name}' ({count}x)")

            print("\n" + "=" * 60)
            print("AUTOMATION COMPLETE - Check screenshots folder")
            print("=" * 60)

            print("\nKeeping browser open 20s for inspection...")
            time.sleep(20)

        except Exception as e:
            print(f"\nFATAL ERROR: {e}")
            ss(page, 'fatal-error')
            import traceback
            traceback.print_exc()
        finally:
            browser.close()


if __name__ == '__main__':
    main()

File: scripts/appsmith-import.js
// Appsmith Import Automation via Playwright
const { chromium } = require('playwright');
const path = require('path');
const fs = require('fs');

const APPSMITH_URL = 'https://appsmith-production-f331.up.railway.app';
const EMAIL = 'ewertonfm00@gmail.com';
const PASSWORD = 'Mll031119';
const JSON_FILE = path.join(__dirname, '../portal/appsmith-import.json');

async function sleep(ms) {
  return new Promise(r => setTimeout(r, ms));
}

async function screenshot(page, name) {
  const dir = path.join(__dirname, 'screenshots');
  if (!fs.existsSync(dir)) fs.mkdirSync(dir, { recursive: true });
  await page.screenshot({ path: `${dir}/${name}.png`, fullPage: true });
  console.log(`Screenshot saved: ${name}.png`);
}

async function main() {
  console.log('Starting Appsmith import automation...');
  console.log(`JSON file: ${JSON_FILE}`);
  console.log(`File exists: ${fs.existsSync(JSON_FILE)}`);

  const browser = await chromium.launch({
    headless: false,  // visible for debugging
    slowMo: 500,
  });

  const context = await browser.newContext({
    viewport: { width: 1440, height: 900 },
  });

  const page = await context.newPage();

  try {
    // Step 1: Navigate to Appsmith
    console.log('\n--- Step 1: Navigate to Appsmith ---');
    await page.goto(APPSMITH_URL, { waitUntil: 'networkidle', timeout: 30000 });
    await screenshot(page, '01-initial-page');
    console.log(`Current URL: ${page.url()}`);

    // Step 2: Login
    console.log('\n--- Step 2: Login ---');

    // Check if we need to login
    if (page.url().includes('/user/login') || page.url().includes('/login')) {
      console.log('On login page, filling credentials...');
    } else {
      // Try to navigate to login
      await page.goto(`${APPSMITH_URL}/user/login`, { waitUntil: 'networkidle', timeout: 30000 });
    }

    await screenshot(page, '02-login-page');

    // Fill email
    const emailInput = await page.locator('input[name="username"], input[type="email"], input[placeholder*="email" i], input[placeholder*="Email" i]').first();
    await emailInput.fill(EMAIL);

    // Fill password
    const passwordInput = await page.locator('input[name="password"], input[type="password"]').first();
    await passwordInput.fill(PASSWORD);

    // Click sign in
    const signInBtn = await page.locator('button[type="submit"], button:has-text("Sign in"), button:has-text("Login"), button:has-text("Log in")').first();
    await signInBtn.click();

    // Wait for navigation after login
    await page.waitForNavigation({ waitUntil: 'networkidle', timeout: 30000 }).catch(() => {});
    await sleep(2000);
    await screenshot(page, '03-after-login');
    console.log(`After login URL: ${page.url()}`);

    // Step 3: Find the workspace/apps page
    console.log('\n--- Step 3: Looking for workspace ---');

    // Wait for apps to load
    await sleep(3000);
    await screenshot(page, '04-workspace');

    // Look for the existing app or import option
    console.log('Page title:', await page.title());

    // Check if there's an import option at workspace level
    // First try to find "Import" button or option
    const importAtWorkspace = await page.locator('button:has-text("Import"), a:has-text("Import"), [data-testid="import"]').count();
    console.log(`Import buttons found at workspace: ${importAtWorkspace}`);

    // Look for any application cards
    const appCards = await page.locator('[data-testid="t--application-card"], .t--application-card, [class*="ApplicationCard"]').count();
    console.log(`App cards found: ${appCards}`);

    await screenshot(page, '05-looking-for-app');

    // Step 4: Try to find the import option
    console.log('\n--- Step 4: Finding import option ---');

    // Strategy 1: Look for "Create New" or "+" button that might have import option
    const createNewBtn = await page.locator('button:has-text("Create New"), button:has-text("New"), [data-testid="create-app"]').first();
    if (await createNewBtn.isVisible().catch(() => false)) {
      console.log('Found Create New button, clicking...');
      await createNewBtn.click();
      await sleep(1500);
      await screenshot(page, '06-create-new-menu');
    }

    // Strategy 2: Look for import from file in dropdown
    const importFromFile = await page.locator('text="Import from file", text="Import", [data-testid="import-app"]').first();
    if (await importFromFile.isVisible().catch(() => false)) {
      console.log('Found Import from file option!');
      await importFromFile.click();
      await sleep(1000);
      await screenshot(page, '07-import-dialog');
    } else {
      console.log('Import from file not visible yet, trying other approaches...');

      // Close any open menus
      await page.keyboard.press('Escape');
      await sleep(500);

      // Try clicking on the "..." menu of an existing app
      const moreOptionsBtn = await page.locator('[data-testid="t--application-card-context-menu"], .t--application-card-context-menu, button[aria-label*="more" i]').first();
      if (await moreOptionsBtn.isVisible().catch(() => false)) {
        console.log('Found more options button on app card');
        await moreOptionsBtn.click();
        await sleep(1000);
        await screenshot(page, '07b-app-context-menu');
      }
    }

    // Look for all visible text on page to understand structure
    const bodyText = await page.evaluate(() => {
      const elements = document.querySelectorAll('button, a, [role="menuitem"]');
      return Array.from(elements).map(el => el.textContent?.trim()).filter(t => t && t.length < 100);
    });
    console.log('Visible interactive elements:', bodyText.slice(0, 30));

    await screenshot(page, '08-current-state');

    // Step 5: Try workspace-level import
    console.log('\n--- Step 5: Attempting workspace import ---');

    // Navigate back to workspace root
    await page.goto(APPSMITH_URL, { waitUntil: 'networkidle', timeout: 30000 });
    await sleep(3000);

    // Look for all buttons and links
    const allBtns = await page.locator('button, [role="button"]').all();
    console.log(`Total buttons found: ${allBtns.length}`);

    for (const btn of allBtns.slice(0, 20)) {
      const text = await btn.textContent().catch(() => '');
      if (text && text.trim()) {
        console.log(`  Button: "${text.trim()}"`);
      }
    }

    await screenshot(page, '09-workspace-final');

    // Try clicking "Import" if found anywhere
    const importBtns = await page.locator('button:has-text("Import"), a:has-text("Import")').all();
    console.log(`\nImport buttons on page: ${importBtns.length}`);

    if (importBtns.length > 0) {
      console.log('Clicking Import button...');
      await importBtns[0].click();
      await sleep(2000);
      await screenshot(page, '10-import-clicked');

      // Look for file input or "Import from file" option
      const fileInput = await page.locator('input[type="file"]').first();
      if (await fileInput.isVisible().catch(() => false)) {
        console.log('Found file input!');
        await fileInput.setInputFiles(JSON_FILE);
        await sleep(2000);
        await screenshot(page, '11-file-selected');
      } else {
        // Look for "Import from file" text
        const importFromFileOpt = await page.locator('text=/import from file/i, text=/from file/i').first();
        if (await importFromFileOpt.isVisible().catch(() => false)) {
          await importFromFileOpt.click();
          await sleep(1500);

          // Now look for file input
          const fileInput2 = await page.locator('input[type="file"]').first();
          if (await fileInput2.isVisible().catch(() => false) || await fileInput2.isAttached().catch(() => false)) {
            await fileInput2.setInputFiles(JSON_FILE);
            await sleep(2000);
            await screenshot(page, '12-file-selected');
          }
        }
      }
    }

    // Step 6: Handle datasource reconnection dialog
    console.log('\n--- Step 6: Handling datasource dialog ---');
    await sleep(3000);
    await screenshot(page, '13-after-import');

    // Look for datasource reconnect dialog
    const datasourceDialog = await page.locator('text=/datasource/i, text=/reconnect/i, text=/ML Portal DB/i').count();
    console.log(`Datasource dialog elements: ${datasourceDialog}`);

    // Look for "Skip" or "Reconnect" buttons
    const skipBtn = await page.locator('button:has-text("Skip"), button:has-text("Later"), button:has-text("Continue")').first();
    const reconnectBtn = await page.locator('button:has-text("Reconnect"), button:has-text("Connect"), button:has-text("Save")').first();

    if (await reconnectBtn.isVisible().catch(() => false)) {
      console.log('Found reconnect button, clicking...');
      await reconnectBtn.click();
      await sleep(2000);
    } else if (await skipBtn.isVisible().catch(() => false)) {
      console.log('Found skip button, clicking...');
      await skipBtn.click();
      await sleep(2000);
    }

    await screenshot(page, '14-final-state');
    console.log(`Final URL: ${page.url()}`);

    console.log('\n=== Automation completed ===');
    console.log('Check screenshots in scripts/screenshots/ for details');

    // Keep browser open for 10 seconds to see final state
    await sleep(10000);

  } catch (error) {
    console.error('Error during automation:', error);
    await screenshot(page, 'error-state');
    throw error;
  } finally {
    await browser.close();
  }
}

main().catch(console.error);
